rfc/20240824-provider-cache-locking.md

```diff
@@ -0,0 +1,68 @@
+Initializing provider plugins...
+- Reusing previous version of hashicorp/local from the dependency lock file
+- Using hashicorp/local v2.5.1 from the shared cache directory
+```
+
+Any number of tofu instances should be able to be run in parallel and still have safe access to the global cache directory.
+
+#### Scenarios which currently cause failures
+
+Multiple init/plans/apply without a primed cache.
+
+One possibility is that each project downloads and overwrites the same provider files.  The global provider cache scan is not a fast process and should be run as little as possible, it is therefore only run at the start of `tofu init`. ProjectA and ProjectB's inits may be run at the same time.  When this occurs, they will both download the full set of providers that they need and clobber each others files.  This is not ideal, but is only a waste of time/resources.
+
+Another possibility is that one project may already be planning/applying and have a live provider executable running, effectively locking it on disk.  If ProjectA is still downloading providers during init and ProjectB has a much smaller list and is already in plan, ProjectA may attempt to overwrite the provider binary that ProjectB is running and maintaining an execution lock on.
+
+
+Mixture of missing platform/corrupt lock files with valid lock files.
+
+The provider lock file may have been generated on a system with a different architecture or may have become corrupt due to a variety of reasons. When this happens, the global cache may not match the lock file and force a re-download of the provider. This will cause potentially unexpected downloads in the above scenarios.
+
+### Technical Approach
+
+This proposal hinges on a stable and consistent cross-platform lock. The codebase already contains this in the form of locking local state files.  This code is battle tested and overall quite simple to use.  With some light refactoring, this filesystem locking code can be moved into it's own internal package and used both to lock providers and to lock the local state file.
+
+The existing file locking should be safe on any local filesystems, but should be used with caution on shared volumes such as legacy NFS shares who do not provide strong locking consistency.
+
+Within the provider installation code, the whole section which inspects and links to the current providers available should be locked at the provider level.  The lock should be done at this granularity specifically for the complicating factors mentioned above.
+
+Additionally, we should make the package installer smarter and able to check the files in the cache against the downloaded version. This prevents a bad cache entry from overwriting a valid entry which may already been in use by another process.
+
+### Open Questions
+
+* Is fnctl flock safe enough? It is industry standard and battle tested.  Are there any additional caveats that should be mentioned in the docs?
```

Comment:
I would love if we could discuss options above. I'm a bit concerned that we've discussed that we do want to lock, but not *how* we want to lock until you mention fnctl here. 

I think this ties into my comment about explaining 'file system level locking' above.

Reply:
ebf7dae4c4